### PR TITLE
Allow to hide version warning

### LINKED ISSUE #1616: How to disable the "You are not using the most up to date version" warning
Thanks for making ReadTheDocs happen first of all. I just discovered that the .com offering launched so you can expect us from [Graylog](http://www.graylog.org/) to upgrade soon.

We are releasing v1.2 of our software as an RC these days. When activating the `1.2` branch in RTD I noticed this warning on the RTD pages of `1.0` and `1.1`:

> You are not using the most up to date version of the library. 1.2 is the newest version.

The current default version is `1.1` so everybody who opens the documentation is seeing that warning. This is misleading because `1.2` is not stable yet.
- Can we disable that warning somehow?
- Can we change the copy/text of the warning? The warning makes sense once `1.2` is released GA but we are not documenting a _library_.

Thanks again and let me know if you need any more information.

### PR DESCRIPTION
**Pull request by stsewd**
Closes #1616

I added a boolean field to allow users hide the version warning (_You are not using..._). This option is per project, no per version.

This is a initial proposal, let me know if you are interested on adding this option.

Things that are missing:

- Run migrations
- Verify if another place needs this field (other endpoints of the api?)
- Add/update tests?
- Update documentation

Related feature #3547

### CODE REVIEW

**Comment by humitos** [comment]
I think the implementation that you propose here does the trick on disabling the warning, but is that what we want? For example, setting the `Default branch/tag` shouldn't be enough for the original use case proposed in that issue. I mean, if the user set `1.0` as default branch maybe,

* when you hit `1.0` it should say nothing
* on `1.1` and `1.2` it could say "This version is not released yet"

I don't know, I'm just thinking aloud. I know there are other issues related to this one that we may consider to find the general solution instead of a specific one as it is this one. We can continue discussing. For example #3547 

Besides, in the issue there are other considerations like i18n and the text from the note itself.

Finally, in case that we continue this direction, this shouldn't be added to the web interface but to the yaml as a config since the web interface it's kind on a deprecation status.

**Comment by stsewd** [comment]
@humitos yeah, you are right, #3547 is a better solution for me. But allowing to users to choosing the _level of integration_ of rtd could be some step here, anyway, just saying, I'm not sure if that would be a real case.

**Comment by humitos** [comment]
I'm marking this as `Needed: design decision` since I think this is not the direction we want to go but I will leave this opened so @agjohnson or @ericholscher can make an opinion here also.

**Comment by ericholscher** [comment]
A couple things:

* I'm generally -1 on adding new settings unless it's really important. Is this something we really need, or should we make the feature work better?
* We should be adding new settings to the YAML file, not to our database. Of course, we need to keep some state in the DB around the YAML files, so that is a larger project, with more design decisions...

**Comment by stsewd** [comment]
> Is this something we really need, or should we make the feature work better?

I think #3547 is a better improvement.

So, should I close this PR? And also we should close the related issue and linking to #3547

**Comment by humitos** [comment]
I wrote a comment at https://github.com/rtfd/readthedocs.org/issues/3481#issuecomment-378000845 that, if implemented, will cover this case.

**Comment by stsewd** [comment]
Closing then

**Comment by humitos** [comment]
We had a discussion yesterday with the core team and we made a decision regarding this topic. Since we need a way to "solve" the problem that some project doesn't match our rules to decide whether to show the banner or not and we can't make those rules extemedely flexible, we decided:

* save this setting in the Project db model
* mark the option as default `False` (so, people needs to enable it if they want to show the banner)
* create a migration that set this field in True to existing projects

Basically, this PR almost fits all the needings here and we will want to continue working on this.

@stsewd would you like to wrap this PR and make it ready to be merged?

**Comment by stsewd** [comment]
@humitos sure

> people needs to enable it if they want to show the banner

I'm don't understand this, isn't this should be the default behavior? I mean, this is one of the top features of rtd (at least I think so).

**Comment by humitos** [comment]
> I'm don't understand this, isn't this should be the default behavior? I mean, this is one of the top features of rtd (at least I think so).

Yes, it's a really good feature... when it works :)

I mean, if the project's release rules matches our logic for the banner, it's a really good feature. Otherwise, it's annoying.

So, the default value is arguable depending on the answer to "how many project we have where this is a feature" over "how many project we have where this is a problem", I'd say. Now, how do we get those numbers?

I think the idea of making it default to `False` it's because, if your project doesn't follow our rules, it's broken by default. Which is not a good experience.

**Comment by stsewd** [comment]
> create a migration that set this field in True to existing projects

I'm not sure how to do that, this is done when applying the migration on the rtd servers?

We don't have docs about this feature, so we can add them now, maybe here https://docs.readthedocs.io/en/latest/versions.html?

**Comment by humitos** [comment]
Yes to both.

For the migration, take a look at the django docs (I'm in the phone now) to see how to create a method that's executed when the migration is ran. You can test this locally in your instance first.

For the docs, we should add it, yes.

**Comment by stsewd** [comment]
@humitos I was searching for those docs, I didn't find then maybe I even didn't know what was searching :grin:.

**Comment by stsewd** [comment]
This is ready for review, I tested the data migration and works perfectly :)

File: docs/versions.rst [hunk 1]
```diff
@@ -55,3 +55,13 @@ for example ``http://pip.readthedocs.io/``,
 they will be redirected to the **Default version**.
 This defaults to **latest**,
 but could also point to your latest released version.
+
+Version warning
+---------------
+
+This is the banner with the text *You are not using..*
+that is displayed on an unreleased version of your docs,
```
[the remark this answers is not in the document]
Comment (stsewd) [reply docs/versions.rst]: Yeah, my mistake

File: readthedocs/projects/migrations/0025_show-version-warning-existing-projects.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.9.13 on 2018-05-07 19:25
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+def show_version_warning_to_existing_projects(apps, schema_editor):
+    Project = apps.get_model('projects', 'Project')
+    for project in Project.objects.all():
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/migrations/0025_show-version-warning-existing-projects.py]: I was wondering how much time this would take on the rtd servers p:

File: readthedocs/projects/migrations/0024_add-show-version-warning.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    dependencies = [
+        ('projects', '0023_migrate-alias-slug'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='project',
+            name='show_version_warning',
+            field=models.BooleanField(default=False, help_text='Show warning banner in non-stable versions.', verbose_name='Show version warning'),
```
Comment (stsewd) [inline readthedocs/projects/migrations/0024_add-show-version-warning.py]: > in non-stable versions

I think this should be different since the banner isn't shown in the latest version also

File: readthedocs/projects/migrations/0024_add-show-version-warning.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        ),
+        migrations.AlterField(
+            model_name='project',
+            name='python_interpreter',
+            field=models.CharField(choices=[('python', 'CPython 2.x'), ('python3', 'CPython 3.x')], default='python', help_text='(Beta) The Python interpreter used to create the virtual environment.', max_length=20, verbose_name='Python Interpreter'),
+        ),
+        migrations.AlterField(
+            model_name='project',
+            name='repo',
+            field=models.CharField(help_text='Hosted documentation repository URL', max_length=255, validators=[readthedocs.core.validators.RepositoryURLValidator()], verbose_name='Repository URL'),
```
Comment (stsewd) [inline readthedocs/projects/migrations/0024_add-show-version-warning.py]: Here a validator was added

File: readthedocs/projects/migrations/0024_add-show-version-warning.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            field=models.IntegerField(default=0, help_text='Number of times this domain has been hit'),
+        ),
+        migrations.AlterField(
+            model_name='project',
+            name='allow_promos',
+            field=models.BooleanField(default=True, help_text='If unchecked, users will still see community ads.', verbose_name='Allow paid advertising'),
+        ),
+        migrations.AlterField(
+            model_name='project',
+            name='comment_moderation',
```
Comment (stsewd) [inline readthedocs/projects/migrations/0024_add-show-version-warning.py]: This needs deletion

File: readthedocs/projects/migrations/0024_add-show-version-warning.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: Just found two things that are commented below, the rest are just help text updates

**Comment by stsewd** [comment]
We need to reminified the js here too, but I'm not sure if you want me to add this or maybe you want to do this from your side :)

File: docs/versions.rst [hunk 6]
```diff
@@ -55,3 +55,13 @@ for example ``http://pip.readthedocs.io/``,
 they will be redirected to the **Default version**.
 This defaults to **latest**,
 but could also point to your latest released version.
+
+Version warning
+---------------
+
+This is the banner with the text *You are not using..*
+that is displayed on older versions of your docs (not stable or latest),
+redirecting the users to the latest version of your docs.
```
Comment (humitos) [inline docs/versions.rst]: Now, thinking a little more on this, I think we need to be more clear here on what is this exactly since it's disabled by default. Consider that the user maybe haven't seens this message before and we need to communicate what it this banner, how it will looks like and when it will appear.

Considering those points, could you expand these docs a little more?

File: readthedocs/projects/migrations/0024_add-show-version-warning.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: Thanks for checking this.

BTW, I saw only one comment. The one about the validator. Which is the other one?

Comment (humitos) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: Yes, agree. Please, fix this in the model and here.

[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: https://github.com/rtfd/readthedocs.org/pull/3595#discussion_r186923805

**Comment by stsewd** [comment]
> Since we don't modified any form or template to show the new field under the admin page

Actually, there is a modification on a form, and yes, I tested it, works as expected.

File: readthedocs/projects/migrations/0024_add-show-version-warning.py (hunk 3 above)
Comment (stsewd) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: Any suggestions? Maybe Show warning banner in old versions?

**Comment by humitos** [comment]
> Actually, there is a modification on a form

Ups... Sorry, I missed that change.

File: readthedocs/projects/migrations/0024_add-show-version-warning.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: I liked the "non-stable" since it's clear. Maybe:

> Show warning banner in non-stable nor latest versions

What do you think?

Comment (stsewd) [reply readthedocs/projects/migrations/0024_add-show-version-warning.py]: I wanna to avoid the `nor`, but it's better than old versions

**Comment by ericholscher** [comment]
Would be good to get this merged. It looks good to me, but I haven't been following the PR convo. @humitos you good with this going in?

**Comment by humitos** [comment]
This is ready to be merge from my point of view.

We will need to be careful when doing the deploy since we are going to be updating **ALL** the projects.

So, before merging, maybe it worth that @agjohnson takes a quick look at it. Although, I don't want to keep blocking this for too much time.

**Comment by stsewd** [comment]
I just rebase and rebuild the assets

**Comment by ericholscher** [comment]
Merging this. Need to do the migration next week, but will be good to get this out. 🍾

**Comment by astrofrog** [comment]
Is there any way to set this setting in the ``readthedocs.yml`` branch so that different branches/tags could have a different setting?

**Comment by humitos** [comment]
@astrofrog let me know if you have any doubt with https://github.com/humitos/sphinx-version-warning or if it does not suit your needs. There are some examples and it's easy to setup and customize.
